pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,343 @@
+package controllers
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kapierrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/credentialprovider"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/secret"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/golang/glog"
+)
+
+const ServiceAccountTokenSecretNameKey = "openshift.io/token-secret.name"
+const OpenshiftDockerURL = "docker-registry.default.svc.cluster.local"
+
+// DockercfgControllerOptions contains options for the DockercfgController
+type DockercfgControllerOptions struct {
+	// ServiceAccountResync is the time.Duration at which to fully re-list service accounts.
+	// If zero, re-list will be delayed as long as possible
+	ServiceAccountResync time.Duration
+}
+
+// NewDockercfgController returns a new *DockercfgController.
+func NewDockercfgController(cl client.Interface, options DockercfgControllerOptions) *DockercfgController {
+	e := &DockercfgController{
+		client: cl,
+	}
+
+	_, e.serviceAccountController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.ServiceAccount{},
+		options.ServiceAccountResync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAccountAdded,
+			UpdateFunc: e.serviceAccountUpdated,
+			DeleteFunc: e.serviceAccountDeleted,
+		},
+	)
+
+	return e
+}
+
+// DockercfgController manages ServiceAccountToken secrets for ServiceAccount objects
```

Comment:
these files are probably the place I need comments the most to understand what each piece is responsible for